cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -477,13 +477,19 @@ const std::pair<bool, int64_t> kTimeConversionTable[4][4] = {
 
 }  // namespace
 
-template <>
-struct CastFunctor<TimestampType, TimestampType> {
+// <TimestampType, TimestampType> and <DurationType, DurationType>
+template <typename O, typename I>
+struct CastFunctor<
+    O, I,
+    typename std::enable_if<(std::is_base_of<O, TimestampType>::value &&
+                             std::is_base_of<I, TimestampType>::value) ||
+                            (std::is_base_of<O, DurationType>::value &&
+                             std::is_base_of<I, DurationType>::value)>::type> {
   void operator()(FunctionContext* ctx, const CastOptions& options,
```

Comment:
So for some reason, this works for `<TimestampType, TimestampType>`, but not for `<DurationType, DurationType>`. 
This functions calls `ShiftTime<int64_t, int64_t>` (defined just above), and in there, it is the line

```
auto out_data = output->GetMutableValues<out_type>(1);
```

that segfaults. But I would think that TimestampType and DurationType have the exact same layout, and should not behave differently?

Reply:
Is that line crashes, then it probably means `output` is null.

Reply:
So, the issue is probably that `is_primitive` (in `arrow/type_traits.h`) returns false for Duration. It should probably be fixed.

Reply:
Thanks, that indeed seems to fix it